### PR TITLE
[Enhancement] Optimize Gson subtype  deserialize

### PR DESCRIPTION
**Pull request by gengjun-git**
Currently, the search of subType in the gson util when deserializing is as follows:
1. if the typeFieldName(we use `clazz`) is specified in the json string, then use this class.
2. if default class if specified when register, then use the default class.
3. throw exception.

But we should support this case: there is no `clazz` field nor specified default class, but we use a subclass to deserialize the json string.
To support this case, we add another search logic:
 - if the class specified (the class is the second param of `fromJson`) is in the register list, then use this class.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/persist/gson/RuntimeTypeAdapterFactory.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/persist/gson/RuntimeTypeAdapterFactory.java|2|2|100.00%|[]|

File: fe/fe-core/src/test/java/com/starrocks/persist/gson/RuntimeTypeAdapterFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (imay) [inline fe/fe-core/src/test/java/com/starrocks/persist/gson/RuntimeTypeAdapterFactoryTest.java:1]: Should use StarRocks license header.

File: fe/fe-core/src/main/java/com/starrocks/persist/gson/RuntimeTypeAdapterFactory.java [hunk 2]
```diff
@@ -361,11 +361,17 @@ public R read(JsonReader in) throws IOException {
...
+                //1. if the typeFieldName is specified in the json string, then use this class.
+                //2. if default class if specified, then use the default class.
+                //3. if the type specified (the second param of GSON.fromJson) is in the register list,
+                //   then use this class.
                 String label;
                 if (labelJsonElement != null) {
                     label = labelJsonElement.getAsString();
                 } else if (defaultLabel != null) {
                     label = defaultLabel;
+                } else if (subtypeToLabel.containsKey(type.getRawType())) {
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/persist/gson/RuntimeTypeAdapterFactory.java:373]: you should also remove `.registerSubtype(InsertTxnCommitAttachment.class, InsertTxnCommitAttachment.class.getSimpleName(), true)` in this pr https://github.com/StarRocks/starrocks/pull/23833/files ?

File: fe/fe-core/src/test/java/com/starrocks/persist/gson/RuntimeTypeAdapterFactoryTest.java (hunk 1 above)
Comment (gengjun-git) [reply fe/fe-core/src/test/java/com/starrocks/persist/gson/RuntimeTypeAdapterFactoryTest.java]: done

File: fe/fe-core/src/main/java/com/starrocks/persist/gson/RuntimeTypeAdapterFactory.java (hunk 2 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/persist/gson/RuntimeTypeAdapterFactory.java:373]: done

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/persist/gson/RuntimeTypeAdapterFactory.java|2|2|100.00%|[]|

**Review by nshangyiming — APPROVED** [review]

**Review by Astralidea — APPROVED** [review]

**Pull request merged**